### PR TITLE
Support loading ES Modules

### PR DESCRIPTION
**Pull request by beerose**
Continuation of https://github.com/cosmiconfig/cosmiconfig/pull/283

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -86,6 +81,7 @@ const defaultLoaders = Object.freeze({
   noExt: loaders.loadYaml,
 } as const);
 const defaultLoadersSync = Object.freeze({
+  '.mjs': loaders.loadJsSync,
```
Comment (d-fischer) [inline src/index.ts:84]: I don't think that works. ESM can only load async.

File: src/index.ts [hunk 2]
```diff
@@ -195,11 +185,15 @@ function cosmiconfigSync(
   };
 }
 
+function normalizeOptions(
+  moduleName: string,
+  options: Options,
+): ExplorerOptions;
```
Comment (d-fischer) [inline src/index.ts]: This overload seems useless, since it has the same signature as the implementation.

**Review by d-fischer — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 2 above)
Comment (beerose) [reply src/index.ts]: Yes, it's still a draft. Working on it

**Comment by tianyingchun** [comment]
it seems that this change aim to `native esm` with `.mjs` 

but while our `pckage.json` declare`type:module` with source file extention `.ts`,`.js`, it will as `esm` also.

**Comment by beerose** [comment]
> it seems that this change aim to `native esm` with `.mjs`
> 
> but while our `pckage.json` declare`type:module` with source file extention `.ts`,`.js`, it will as `esm` also.

AFAIK that's something that typescript loader will take care of. @d-fischer can you confirm?

File: src/index.ts (hunk 1 above)
Comment (beerose) [reply src/index.ts]: Yep, made a mistake when solving conflicts. Updated now

File: package.json [hunk 3]
```diff
@@ -92,10 +89,10 @@
...
   },
   "devDependencies": {
-    "@babel/cli": "^7.10.4",
-    "@babel/core": "^7.10.4",
-    "@babel/preset-env": "^7.10.4",
-    "@babel/preset-typescript": "^7.10.4",
+    "@babel/cli": "7.19.3",
+    "@babel/core": "7.20.2",
+    "@babel/preset-env": "7.20.2",
+    "@babel/preset-typescript": "7.18.6",
```
Comment (d-fischer) [inline package.json:95]: Any reason why this switches to exact versioning?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## 8.1.4
+
+- **Breaking change:** Add support for ECMAScript modules (ESM) to the [*asynchronous* API](./README.md#asynchronous-api). End users running Node versions that support ESM can provide `.mjs` files, or `.js` files whose nearest parent `package.json` file contains `"type": "module"`.
```
Comment (d-fischer) [inline CHANGELOG.md:5]: Is this really a breaking change?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## 8.1.4
```
Comment (d-fischer) [inline CHANGELOG.md:3]: Should be added as `HEAD`. Versioning concerns are not PR concerns.

File: package.json [hunk 6]
```diff
@@ -25,12 +25,6 @@
     "check:all": "npm run test && npm run typescript && npm run lint && npm run format:check",
     "prepublishOnly": "npm run check:all && npm run build"
   },
-  "husky": {
-    "hooks": {
-      "pre-commit": "lint-staged && npm run typescript && npm run test",
-      "pre-push": "npm run check:all"
-    }
-  },
```
Comment (d-fischer) [inline package.json:33]: Where did these go?

**Review by d-fischer — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 7]
```diff
@@ -91,33 +101,25 @@ function replaceMetaPlaceholders(
...
+  options: T,
+): T {
   const metaExplorer = new ExplorerSync({
     packageProp: 'cosmiconfig',
     stopDir: process.cwd(),
     searchPlaces: metaSearchPlaces,
     ignoreEmptySearchPlaces: false,
     usePackagePropInConfigFiles: true,
-    loaders: defaultLoaders,
+    loaders: { ...defaultLoadersSync, ...defaultLoaders },
```
Comment (d-fischer) [inline src/index.ts:114]: We can't really add async loaders here since it's an inherently sync explorer.

File: package.json (hunk 6 above)
Comment (beerose) [reply package.json:33]: I think they were removed in the previous esm-support PR (the one I'm building on top of). Will add it back

File: package.json (hunk 3 above)
Comment (beerose) [reply package.json:95]: I suppose not.

File: CHANGELOG.md (hunk 5 above)
Comment (beerose) [reply CHANGELOG.md:3]: Can you elaborate? What should I change it to?

Comment (d-fischer) [reply CHANGELOG.md:3]: It should literally just be `HEAD`, as previous PRs have done it as well. Assigning a real version to this is part of the release procedure.

Comment (d-fischer) [reply CHANGELOG.md]: No you did not get it. This is not part of 8.1.3, which is already released. The convention is to literally write "HEAD" as the version.

Comment (beerose) [reply CHANGELOG.md]: Done.

**Review by d-fischer — APPROVED** [review]

**Comment by tianyingchun** [comment]
it seems has a bug here 
```ts
const loadJs: LoaderAsync = async function loadJs(filepath) {
  try {
    const { href } = pathToFileURL(filepath);
    return (await import(href)).default;
  } catch (error) {
    return loadJsSync(filepath, null);
  }
};
```
if we have named export here it will always return `undefined`

maybe it should be changed to 
```ts
const { href } = pathToFileURL(filepath);
const moduleData =  await import(href);
return moduleData.default.|| moduleData;
```

**Comment by d-fischer** [comment]
I don't see why it should load a non-default export from the file. Do you have any actual use case for this? I'm pretty sure that vite does not use cosmiconfig.

**Pull request merged**